src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,191 @@
+
+    private int size = 0;
+    private Node<T> head;
+    private Node<T> tail;
+
+    private static class Node<T> {
+        private T item;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(Node<T> prev, T element, Node<T> next) {
+            this.item = element;
+            this.next = next;
+            this.prev = prev;
+        }
+    }
+
+    private void validIndexCheck(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Invalid index: " + index);
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> node = new Node<>(null, value, null);
+        if (size == 0) {
+            head = node;
+            tail = head;
+        } else {
+            node = new Node<>(tail,value,null);
+            if (tail.prev == null) {
+                tail.prev = head;
+                head.next = node;
+            } else {
+                tail.next = node;
+                node.prev = tail;
+            }
+            tail = node;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index != size) {
+            validIndexCheck(index);
+        }
+        if (index == 0 && size != 0) {
+            Node node = new Node(null, value, head);
+            head.prev = node;
+            node.next = head;
+            head = node;
+            size++;
+        } else {
+            if (index == size || size == 0) {
+                add(value);
+            } else {
+                Node node = new Node(head, value, head.next);
+                for (int i = 1; i < index; i++) {
+                    node.prev = node.next;
+                    node.next = node.next.next;
+                }
+                node.next.prev = node;
+                node.prev.next = node;
+                size++;
+            }
+        }
+
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        validIndexCheck(index);
+        Node<T> node = head;
+        for (int i = 0; i < index; i++) {
+            node = node.next;
+        }
+        return node.item;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        validIndexCheck(index);
+        T oldValue;
+        if (index == 0) {
+            oldValue = head.item;
+            head.item = value;
+            return oldValue;
+        }
+        if (index == size - 1) {
+            oldValue = tail.item;
+            tail.item = value;
+        } else {
+            Node node = new Node(head, value, head.next);
+            for (int i = 1; i < index; i++) {
+                node.prev = node.next;
+                node.next = node.next.next;
+            }
+            oldValue = (T) node.next.item;
+            node.next.item = value;
```

Comment:
The logic in the set method is incorrect. You should not create a new node here. Instead, find the existing node at the given index and update its value. The loop logic is also incorrect as it does not correctly traverse the nodes.